### PR TITLE
Implement FACTS devices SVC and TCSC

### PR DESCRIPTION
**Pull request by rbolgaryn**
We follow the PhD Thesis of Ara Panosyan to implement Static Var Compensator (SVC) and Thyristor-Controlled Series Capacitor (TCSC).

The SVC device controls bus voltage magnitude with a local set-point at its connection bus via adjusting the shunt impedance of the element. 

The electrical scheme bnehind the SVC element is as shown below:

svc

The fixed capacitor is represented by the parameter x_cvar_ohm. The recator with reactance x_l_ohm is adjustable via a thyristor circuit. The thyristor firing angle effectivbely controls the shunt impedance of the element.

The TCSC controls the active power flow through its path with a local active power flow set-point at its "to" bus via adjusting the series impedance of the element. TCSC element is considered ideal (no real power losses).

The circuit is shown below:

tcsc

The circuit of TCSC is (similarly to SVC) a combination of a capacitor and a reactor. The series impedance of the element is controlled by thyristor firing angle.

The power flow calculation was expanded by:
* Admittance matrices of SVC and TCSC
* Jacobian modification matrices for SVC and TCSC
* Mismatch equations for SVC and TCSC

pandapower library was expanded to include two new elements, net.svc and net.tcsc.
The elements can be "controllable" - actively adjusting the value of thyristor firing angle to reach the set-point. Alternatively, controllable can be set to False and then the elements are equivalent to shunt or impedance element with the impedance value defined by the value of thyristor_firing_angle in net.svc or net.tcsc.

The elements svc and tcsc are represented by separate arrays in ppc: ppc["svc"] and ppc["tcsc"]. 
This required numerous changes in pandapower. 

The new functionality is covered by tests.

The limitations are:
* SVC cannot be attached to a "PV" bus
* Multiple TCSC elements cannot be attached to the same from_bus or to_bus (including if separate buses are connected by switches)

### CODE REVIEW

File: doc/elements/tcsc.rst [hunk 1]
```diff
@@ -0,0 +1,87 @@
+﻿.. _tcsc:
+
+=============
+Thyristor-Controlled Series Capacitor (TCSC)
+=============
+
+We implement the FACTS devices based on the PhD Thesis of Ara Panosyan, PhD.
```
Comment (SteffenMeinecke) [inline doc/elements/tcsc.rst:7]: is there any link or name of the PhD Thesis? the reference is poor at the moment.

File: doc/elements/svc.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
+﻿.. _svc:
+
+=============
+Static Var Compensator (SVC)
+=============
+
+We implement the FACTS devices based on the PhD Thesis of Ara Panosyan, PhD.
```
Comment (SteffenMeinecke) [inline doc/elements/svc.rst:7]: Reference improvements as in tcsc.rst ?

File: pandapower/pf/create_jacobian_facts.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2016-2022 by University of Kassel and Fraunhofer Institute for Energy Economics
```
Comment (SteffenMeinecke) [inline pandapower/pf/create_jacobian_facts.py:3]: 2023 would be correct.
2022 also occurs in two protection files. @rbolgaryn can you fix this at the same time?

File: pandapower/pf/makeYbus_facts.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import numpy as np
```
Comment (SteffenMeinecke) [inline pandapower/pf/makeYbus_facts.py:1]: missing copyright preamble?

File: pandapower/results_branch.py [hunk 5]
```diff
@@ -578,6 +581,67 @@ def _get_impedance_results(net, ppc, i_ft, suffix=None):
...
+        i_to_ka = ppc["tcsc"][f:t, TCSC_IT].real
+        # losses will be 0 because tcsc represents an ideal device but we can add resistance easily in the future
+        pl_mw = p_from_mw + p_to_mw
+        ql_mvar = q_from_mvar + q_to_mvar
+    else:
+        raise NotImplementedError("TCSC not implemented for algorithm != 'ac' - writing zeros to tcsc results. "
+                                  "Results for other elements are as if TCSC elements were open switches.")
+        zeros = np.zeros(t)
+        # this looks like a pyramid
+        p_from_mw = zeros
```
Comment (SteffenMeinecke) [inline pandapower/results_branch.py:614]: A note just to force a check by @rbolgaryn : Might this reusage of the same variable `zeros` lead to undesired behavior?

File: pandapower/test/loadflow/test_facts.py [hunk 6]
```diff
@@ -0,0 +1,608 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2016-2022 by University of Kassel and Fraunhofer Institute for Energy Economics
```
Comment (SteffenMeinecke) [inline pandapower/test/loadflow/test_facts.py:3]: again copyright

File: pandapower/topology/graph_searches.py [hunk 7]
```diff
@@ -120,8 +120,7 @@ def calc_distance_to_bus(net, bus, respect_switches=True, nogobuses=None,
          dist = top.calc_distance_to_bus(net, 5)
 
     """
-    g = create_nxgraph(net, respect_switches=respect_switches,
```
Comment (SteffenMeinecke) [inline pandapower/topology/graph_searches.py:123]: This change violate the 100 chars per line PeP8 limit

File: tutorials/FACTS.ipynb [hunk 8]
```diff
@@ -0,0 +1,756 @@
...
+   "source": [
+    "# FACTS devices"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "167787e5",
+   "metadata": {},
+   "source": [
+    "We implement the FACTS devices Static Var Compensator (SVC) and Thyristor-Controlled Series Capacitor (TCSC) based on the PhD Thesis of Ara Panosyan, PhD.\n",
```
Comment (SteffenMeinecke) [inline tutorials/FACTS.ipynb:16]: same poor reference

File: tutorials/FACTS.ipynb [hunk 9]
```diff
@@ -0,0 +1,756 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "25c2a3a6",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#### define and prepare the generation and loads profiles.\n",
```
Comment (SteffenMeinecke) [inline tutorials/FACTS.ipynb:70]: `loads profiles` -> `load profiles`?

File: tutorials/FACTS.ipynb [hunk 10]
```diff
@@ -0,0 +1,756 @@
...
...icu89WLWq/v01hWSjYdaqXWVtmoefIwoFFOaUUpBTgm33bgAUHj9R7XiDTkfp5Tix3UakGrq0squ2H4BfJzdm/mcQ098fQHBY62r3FjejYZaTUohOp8fa37+il1lu2jXKjI7CWOTgXCFAUE5poWjyXN5Q83pNOvs8Co7eUJQJF/6qe9+aCDdGbbpPkZ/lJkDOTS2EzhNEPZhaAxf+hNWzap7rruX4L+L7XFYg6t6mr5ZG2S1GxH6RxhjcbxBqKxPLKGpBoVBURM3KVR6bPSq1SAFWKEVq5Pte8GUI/HSbqJNOjxRpxNOWCCVRGfltltTbMAsLC2Pjxo0VP5dPePPmzWPAgAHmG5mkEqVSpILM3CB6WWTHid4W/3aD/2fvvMPbKq8//rnakiXvvWInzt5kskISQggjEHYZZZS2QCEt0NJCB1AoBCi7/KBlb1r2JhBIyB5kOns4sR3vLVl73d8fryTbsR3biTxi6/M8fuLo3vveV4l0z3vec873PJwJT42Bp8fCwxnii/jUKCFf/MVvRWrhzo9FD7Avb4fXz4Onx4t0j3Bx8EexuykpRSrKse6+6KKFcwZCActWLZop/npZpLbsOCjcKv6vE60O4q45St55B1QGFBlTcpoWO2qtkthU4SRVFzdimDIVAPtPP7W63lNejux2I6nVWFxC5SoutW3HDEAXpUapavsRZYzTotYq8ftlzFUOlPHx6GXxLLIUF7U8+fTfCxng4jVi4ea2iw2M/10d6BMWA5e/JYRCjoZCKTZJQCizdQVrtai/AREViBB2IrapH5F3pmgToVCL9N/PF4Lf3/a5fp+IYn++UEQCxl8JV38Qka8fYPi8HvavWw0cfxpjkOGnCsesaPtW7OaGsIzZ7WScJFLvJaUQvaovFC1fZD+M+xncur7rGU0RepQuV9YuWrSIefPmsWvXLrxeL8888ww7d+5k7dq1LF9+guwqnKjknAa3rBayqFveEUbI3Sh+miMpRX1X4lDR+DImQ/TKqNkPlTugsVw4bb/68diLq4P4PPBNQC1yyi9Fn4vjYdpNQuwDGUacLxy9o0mcRzgq5qpKzNWVSLJMWmY22sHHpiDmcfuoLRXRptTBLXehk7KN1JfbqC5uJHWqUCx0bNqE7PUiNVPFCyoyqgdlY65yAO1HzDpCkiTi0qKoKrRQX24jPi0Kg8kE+GksL2t5ctp
```
Comment (SteffenMeinecke) [inline tutorials/FACTS.ipynb:730]: the complete tutorial is without any comment what can be interpreted from the calculated and plotted results. Eventhough it is time consuming and information can be found in the PhD, a few short comments would be great, I think.

**Review by SteffenMeinecke — COMMENTED** [review]
I didn't checked calculating facts but only watched at the code changed. Therein, my comments can be found.

**Review by SteffenMeinecke — DISMISSED** [review]

**Review by jwiemer112 — APPROVED** [review]

**Pull request merged**